packages/sources/uniswap-v2/src/endpoint/crypto.ts

```diff
@@ -0,0 +1,122 @@
+
+export const supportedEndpoints = ['crypto']
+
+export const endpointResultPaths = {
+  crypto: 'rate',
+}
+
+export interface ResponseSchema {
+  input: string
+  inputToken: string
+  inputDecimals: number
+  output: string
+  outputToken: string
+  outputDecimals: number
+  rate: number
+}
+
+export const inputParameters: InputParameters = {
+  from: ['base', 'from', 'coin'],
+  fromAddress: false,
+  fromDecimals: false,
+  to: ['quote', 'to', 'market'],
+  toAddress: false,
+  toDecimals: false,
+  amount: false,
+  resultPath: false,
+}
+
+export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
+  const validator = new Validator(request, inputParameters)
+  if (validator.error) throw validator.error
+
+  const jobRunID = validator.validated.id
+  const { address: from, decimals: fromDecimals } = await getTokenDetails(validator, 'from', config)
+  const { address: to, decimals: toDecimals } = await getTokenDetails(validator, 'to', config)
+  const inputAmount = validator.validated.data.amount || 1
+  const amount = BigNumber.from(inputAmount).mul(BigNumber.from(10).pow(fromDecimals))
+  const resultPath = validator.validated.data.resultPath
+
+  const [_amountIn, output] = await getBestRate(from, to, amount, config)
+
+  const outputAmount = new Decimal(output.toString()).div(new Decimal(10).pow(toDecimals))
+  const rate = outputAmount.div(inputAmount)
+
+  const data: ResponseSchema = {
+    input: amount.toString(),
+    inputToken: from,
+    inputDecimals: fromDecimals,
+    output: output.toString(),
+    outputToken: to,
+    outputDecimals: toDecimals,
+    rate: rate.toNumber(),
+  }
+
+  const response = {
+    status: 200,
+    statusText: 'OK',
+    headers: {},
+    config: {},
+    data: data,
+  }
+  const result = Requester.validateResultNumber(response.data, [resultPath])
+
+  return Requester.success(jobRunID, Requester.withResult(response, result), config.verbose)
+}
+
+/**
+ * getTokenDetails will find the address and number of decimal for a token.
+ *
+ * The order of operations is as follows:
+ *  - address:
+ *     1. Check if the address was provided in the request.
+ *     2. If not, check the symbol in the request to see if we have pre-set the address for this symbol/network.
+ *     3. If not, we assume the symbol param was actually the address.
+ *  - decimals:
+ *     1. Check if the number of decimals was provided in the request.
+ *     2. Query the contract at the address found above to see how many decimals it's set to.
+ * @param validator The validation class to use
+ * @param direction Used to get the params in the request
+ * - `{direction}` is the symbol of the token (to find pre-set token details)
+ * - `{direction}Address` is the token address as set in the request
+ * - `{direction}Decimals` is the number of decimals for the token as set in the request
+ * @param config Configuration to extract token decimals from
+ */
+const getTokenDetails = async (
```

Comment:
Is this the exact same function as on the Curve adapter? If so, would it make sense to collect these in a utility function?

Reply:
yes it is, and I agree that it would be useful for other DEX adaptors. Where is the best place to move this function?

Reply:
I'm leaning towards the validator: https://github.com/smartcontractkit/external-adapters-js/blob/develop/packages/core/bootstrap/src/lib/external-adapter/validator.ts#L175

@justinkaseman any opinions?

Reply:
that looks like a logical place to me 👍

Reply:
the only potential issue with moving it into `validator.ts` is that we'll need to import `ethers` and add the `erc20.json` abi into the validator. Do we want to couple web3 logic with the Validator?

Possible changes needed in `validator.ts`
```js
import { ethers, BigNumber } from 'ethers'
import { Decimal } from 'decimal.js'

...


  getDecimals = async (address: string, config: Config): Promise<number> =>
  new ethers.Contract(address, erc20ABI, config.provider).decimals()

  /**
   * getTokenDetails will find the address and number of decimal for a token.
   *
   * The order of operations is as follows:
   *  - address:
   *     1. Check if the address was provided in the request.
   *     2. If not, check the symbol in the request to see if we have pre-set the address for this symbol/network.
   *     3. If not, we assume the symbol param was actually the address.
   *  - decimals:
   *     1. Check if the number of decimals was provided in the request.
   *     2. Query the contract at the address found above to see how many decimals it's set to.
   * @param validator The validation class to use
   * @param direction Used to get the params in the request
   * - `{direction}` is the symbol of the token (to find pre-set token details)
   * - `{direction}Address` is the token address as set in the request
   * - `{direction}Decimals` is the number of decimals for the token as set in the request
   * @param config Configuration to extract token decimals from
   */
  getTokenDetails = async (
    adapter: string,
    validator: Validator,
    direction: string,
    config: Config,
  ): Promise<{ address: string; decimals: number }> => {
    const symbol = this.overrideSymbol(
      adapter,
      this.validated.data[direction],
    ) as string
    const address =
      this.validated.data[`${direction}Address`] ||
      this.overrideToken(symbol, config.network) ||
      symbol
    const decimals =
      this.validated.data[`${direction}Decimals`] || (await this.getDecimals(address, config))

    return { address, decimals }
  }
```

Reply:
I see, that also has a dependency on `Config` (which we can move out, but we still need the ethers provider).

This is growing in scope, so I'm OK with just leaving this as-is for now and I'll create a story for us to put more thought into this later.

I'll do a more thorough review tomorrow with A/T